### ISSUE TITLE
JS files relocation inside the web/js directory not BC in 2.3.x

### ISSUE THREAD

**Issue by rbayet** [magento/magento2#19291]
### Summary

The issue https://github.com/magento/magento2/issues/16302 has been fixed both in 2.2.x and 2.3.x lines but only the fix in 2.2.x is BC compatible by providing and "old_path : new_path" mapping in the requirejs-config.js.

### Examples

For instance for the **2.2.x line** for app/code/Magento/Authorizenet/view/frontend/requirejs-config.js 
 the combination of https://github.com/magento/magento2/commit/8653ba1fbe29acc37505b85adce1463d70e9cdf1 and https://github.com/magento/magento2/commit/e0b09fdc6db266015b3296fe8a272db7c3ba7d3e leads to the following changes : 
From 
```
var config = {
    map: {
        '*': {
            transparent: 'Magento_Payment/transparent'
        }
    }
};
```
To
```
var config = {
    map: {
        '*': {
            transparent: 'Magento_Payment/js/transparent',
            'Magento_Payment/transparent': 'Magento_Payment/js/transparent'
        }
    }
};
```
(See https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Authorizenet/view/frontend/requirejs-config.js)

This means any reference to 'Magento_Payment/transparent' in a third party extension's JS component define() dependency will point correctly to 'Magento_Payment/js/transparent'.

On the other hand, **in the 2.3.x line**, the changes are simply (see https://github.com/magento/magento2/commit/0ce439b1ecc2906dd840f37989c7c8ccf84dc6bf)
From 
```
var config = {
    map: {
        '*': {
            transparent: 'Magento_Payment/transparent'
        }
    }
};
```
To
```
var config = {
    map: {
        '*': {
            transparent: 'Magento_Payment/js/transparent'
        }
    }
};
```
(See https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Authorizenet/view/frontend/requirejs-config.js)

This time, any third party extension's JS component define() dependency to 'Magento_Payment/transparent' will fail and break the extension.

### Proposed solution

As the BC solution found for the 2.2.x line is not that intrusive, it could also be adopted for the 2.3.x line, eliminating (in some conditions) the need to address those JS files relocation for third party extension authors.

Regards,
Richard.

**Comment by magento-engcom-team** [comment]
Hi @rbayet. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@rbayet do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by rbayet** [comment]
Sorry, can't edit the previous comment to tick the checkboxes when appropriate.

The issue is purely related on BC for third party extensions (modules, themes), so it *cannot* be reproduced on a vanilla Magento instance alone.

I could indicate steps to reproduce, but it will require the installation of a third party extension.

**Comment by romainruaud** [comment]
Hello @magento-engcom-team 

Just to be clear, this one is a "Developer experience issue", making us unable to provide proper compatibility with Magento 2.2 AND 2.3 on our Elasticsuite extension.

@rbayet already explained well what we are facing here, and I'm sure we are not alone to build JS components based on the JS components that were affected by this decision to move them.

On our specific case, it's due to the `'Magento_Search/form-mini'` that has been moved, but since many other files are concerned, I'm wondering why you decided to not keep the BC version in 2.3 (you introduced it and used it for 2.2.x versions).

Question is : did you squeeze it on purpose (no BC for major version), or is it a mistake ?

We need to know it in order to design our own compatibility (and BC) map for the next version of our extension which is meant to support Magento 2.3. 

Regards

**Comment by magento-engcom-team** [comment]
Hi @magento-engcom-team. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by romainruaud** [comment]
And btw, @rbayet and me would be glad to submit a PR to fix this one if you ask us to.

Regards

**Comment by romainruaud** [comment]
Just to be transparent : the comment I removed was the auto response when assigned. I let you manage the lifecycle/assignment of this issue by yourself.

Regards

**Comment by magento-engcom-team** [comment]
@romainruaud Thank you for verifying the issue.

Unfortunately, not enough information was provided to created internal ticket. Please consider adding the following:
- [x] Add `"Component: "` label(s) to this ticket based on verification result. If uncertain, you may follow the best guess
- [x] Add `"Reproduced on "` label(s) to this ticket based on verification result

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
@romainruaud Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96618` were created

**Comment by DrewML** [comment]
Thanks for the report @rbayet! I have to assume the missing BC path in `2.3.0` was a mistake. Would be happy to accept a PR from you or @romainruaud with the fix.

**Comment by magento-engcom-team** [comment]
Hi @rbayet. Thank you for your report. 
The issue has been fixed in magento/magento2#19583 by @rbayet in 2.3-develop branch
Related commit(s):
 - [f38f09f879a7495772e5a3b2fc2445731ae2cb5b](https://github.com/magento/magento2/commit/f38f09f879a7495772e5a3b2fc2445731ae2cb5b)

The fix will be available with the upcoming 2.3.1 release.
